src/managers/backup.py

```diff
@@ -84,6 +96,92 @@ def create_bucket(self, s3_parameters: S3ConnectionInfo) -> bool:
 
         return True
 
-    def write_test_string(self) -> None:
-        """Write content in the object storage."""
-        self.bucket.put_object(Key="test_file.txt", Body=b"test string")
+    def create_backup(self) -> BackupMetadata:
+        """Create a snapshot with ZooKeeper admin server and stream it to the object storage."""
+        zk_user = "super"
+        zk_pwd = self.state.cluster.internal_user_credentials.get("super", "")
+        date = datetime.now()
+        snapshot_name = f"{date:%Y-%m-%dT%H:%M:%SZ}"
+        snapshot_path = f"{snapshot_name}/snapshot"
+
+        # It is very likely that the file is fully loaded in memory, because the file-like interface is
+        # not seekable, and I have a strong suspicion that boto uses this to figure out if it can
+        # upload in one go or need to use a multipart request.
+        # We cannot be sure because finding this information in boto code base is time consuming.
+        # If this ever become an issue, we can find a workaround by using the 'content-length' header from
+        # the response. Or write to a temp file as a last resort.
+        with httpx.stream(
+            "GET",
+            f"http://localhost:{ADMIN_SERVER_PORT}/commands/snapshot?streaming=true",
+            headers={"Authorization": f"digest {zk_user}:{zk_pwd}"},
+        ) as response:
+
+            response_headers = response.headers
+            quorum_leader_zxid = int(response_headers["last_zxid"], base=16)
+            metadata: BackupMetadata = {
+                "id": snapshot_name,
+                "log-sequence-number": quorum_leader_zxid,
+                "path": snapshot_path,
+            }
+
+            self.bucket.put_object(
+                Key=os.path.join(self.backups_path, snapshot_name, "metadata.yaml"),
+                Body=yaml.dump(metadata, encoding="utf8"),
+            )
+
+            self.bucket.upload_fileobj(
+                _StreamingToFileSyncAdapter(response),  # type: ignore
+                os.path.join(self.backups_path, snapshot_name, "snapshot"),
+            )
+
+        return metadata
+
+    def format_backups_table(
+        self, backup_entries: list[BackupMetadata], title: str = "Backups"
+    ) -> str:
+        """Format backups metadata into a readable table."""
+        table = Table(title=title)
```

Comment:
**gigantic nitpick:** I'm not a huge fan of adding extra dependencies for niche use cases. This can be achieved with the standard lib through [format specifications](https://docs.python.org/3/library/string.html#format-specification-mini-language). Although getting a pretty result might be a rabbit hole into how to use them.

I'm not against keeping the `rich` package. Just know that when the time comes and `rich` has a vulnerability that breaks every single thing charm related I will pull this comment from the depths of github and say _I told you so_ :smiling_imp:

Reply:
No, you are absolutely right to be cautious! At first, I did not even want to add `httpx` to the 4 letters word PR for those reasons.
In this case, though we do increase the attack surface, I feel like this is an ok approach for the following reasons:
- `rich` is mature enough to be vendorised in `pip` itself. Not to say that this scenario will never happen, but we did do some due diligence on this specific package.
- the current, "manual", alternative used in the others charms is less maintainable in my opinion, while offering way fewer possibilities

If that's a big no for you, I think we should discuss it during a daily sync